[Elixir] New concept exercise: maps (#1336)
exercism/v3: languages/elixir/exercises/concept/maps/.docs/hints.md

### General

> - A [map][maps] are a associative data structure of key-value pairs.

Reviewer: ```suggestion
- A [map][maps] is an associative data structure of key-value pairs.
```

- Elixir offers [many useful Map module functions in the standard library][map-module].

### 1. Define a new high score map

- You need to define a [named function][named-function] with 0 parameters.
- It should return an empty [map][maps].

### 2. Add players to the high score map

> - You need to define a [named function][named-function] with 3 parameters.

Reviewer: You could consider not having these "You need to define a function with X parameters" and their "The first argument is a Y" comments as these things are also explained in the instructions. For the basics exercise it definitely made sense to have these hints as students were defining functions for the first time. 

That said, I don't mind these hints, so feel free to leave them in if you think they're useful.

Author: I'm not sure if they would be helpful or not, they seem redundant for me, but I am not looking at the language with fresh eyes.

If I were to remove them, is there a better hint to leave?

Reviewer: > I'm not sure if they would be helpful or not, they seem redundant for me, but I am not looking at the language with fresh eyes.

I agree with this. They made sense for the very first basics exercise, but less so for subsequent exercises.

> If I were to remove them, is there a better hint to leave?

Well first of all, you don't _have_ to have a hint for each task, although one can usually find something to link to or say.  Looking at the example implementation, maybe you could link to the `Map.put` documentation (without mentioning its name) or another resource that explains how to add things to a Map?

Author: Sounds good, I think I have something I like more now to submit for review.

  - The first argument is a [map][map] of scores.
  - The second argument is a name [string][strings].

>   - The third argument is the score [integer value][integers], it is optional, so a [default value][default-arg] should be specified.

Reviewer: ```suggestion
  - The third argument is the score [integer value][integers]. It is optional, so a [default value][default-arg] should be specified.
```

> - The resulting map should be returned

Reviewer: ```suggestion
- The resulting map should be returned.
```

### 3. Remove players from the score map

- You need to define a [named function][named-function] with 2 parameters.
  - The first argument is a [map][map] of scores.

>   - The second argument is a name [string][strings].

Reviewer: Is there a dependency that ensure students know how to work with strings?

Author: yes, there is a string-literal concept earned in the lists exercise to show the basic string type as data.

We are purposely not manipulating strings until after the pattern matching exercises

Reviewer: Okay! Just checking :)

> - The resulting map should be returned

Reviewer: ```suggestion
- The resulting map should be returned.
```

### 4. Reset a player's score

- You need to define a [named function][named-function] with 2 parameters.
  - The first argument is a [map][map] of scores.
  - The second argument is a name [string][strings].
- The resulting map should be returned with the player's score reset to an initial value.

### 5. Update a player's score

- You need to define a [named function][named-function] with 3 parameters.
  - The first argument is a [map][map] of scores.
  - The second argument is a name [string][strings].
  - The third argument is a score [integer value][integers] to be added to the player's score.
- The resulting map should be returned with the player's updated score.

### 6. Get a list of players with scores ordered by player name

- You need to define a [named function][named-function] with 1 parameters.
  - The first argument is a [map][map] of scores.

> - Maps make use of the [Enumerable][enum] protocol.

Reviewer: I initially read this as maps are using the enumerable protocol. The after document describes it a bit better I think, emphasing that maps implement the enumerable protocol and that there the `Enum` module's functions can be used. Maybe do something like:

```suggestion
- The [`Enum` module][enum]'s function can be used on maps.
```

Or something like that.
